cli/data/data.go

```diff
@@ -0,0 +1,275 @@
+		fmt.Println("Failed to open wallet: ", name)
+		os.Exit(1)
+	}
+	return wallet
+}
+func newContractContextWithoutProgramHashes(data signature.SignableData) *contract.ContractContext {
+	return &contract.ContractContext{
+		Data:       data,
+		Codes:      make([][]byte, 1),
+		Parameters: make([][][]byte, 1),
+	}
+}
+func signTransaction(signer *account.Account, tx *transaction.Transaction) error {
+	signature, err := signature.SignBySigner(tx, signer)
+	if err != nil {
+		fmt.Println("SignBySigner failed")
+		return err
+	}
+	transactionContract, err := contract.CreateSignatureContract(signer.PubKey())
+	if err != nil {
+		fmt.Println("CreateSignatureContract failed")
+		return err
+	}
+	transactionContractContext := newContractContextWithoutProgramHashes(tx)
+	if err := transactionContractContext.AddContract(transactionContract, signer.PubKey(), signature); err != nil {
+		fmt.Println("AddContract failed")
+		return err
+	}
+	tx.SetPrograms(transactionContractContext.GetPrograms())
+	return nil
+}
+
+func readData(filepath string) ([]byte, error) {
+	if _, err := os.Stat(filepath); err != nil {
+		fmt.Printf("invalid file path:%s\n", err)
+		return nil, err
+	}
+
+	f, err := os.OpenFile(filepath, os.O_RDONLY, 0664)
+	defer f.Close()
+	if err != nil {
+		fmt.Printf("open file error:%s\n", err)
+		return nil, err
+	}
+	//read file
+	var payload []byte
+	var eof = false
+	for {
+		if eof {
+			break
+		}
+		buf := make([]byte, 1024)
+		nr, err := f.Read(buf[:])
+
+		switch true {
+		case nr < 0:
+			fmt.Fprintf(os.Stderr, "cat: error reading: %s\n", err.Error())
+			return nil, err
+		case nr == 0: // EOF
+			eof = true
+		case nr > 0:
+			payload = append(payload, buf...)
+
+		}
+	}
+	return payload, nil
+}
+
+func dataAction(c *cli.Context) error {
+	if c.NumFlags() == 0 {
+		cli.ShowSubcommandHelp(c)
+		return nil
+	}
+	upload := c.Bool("upload")
+	reg := c.Bool("reg")
+	get := c.Bool("get")
+	cat := c.Bool("cat")
+	if !upload && !reg && !get && !cat {
+		cli.ShowSubcommandHelp(c)
+		return nil
+	}
+
+	var resp []byte
+	//var txHex string
+	var err error
+	if upload {
+		filepath := c.String("file")
+
+		if filepath == "" {
+			cli.ShowSubcommandHelp(c)
+			return nil
+		}
+
+		payload, err := readData(filepath)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return err
+		}
+
+		fmt.Println("data uploading...")
+		//tranfer data to node
+		resp, err = httpjsonrpc.Call(Address(), "uploadDataFile", 0, []interface{}{payload})
+
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return err
+		}
+		fmt.Println("data uploaded")
+	}
+	if reg {
+		address := c.String("address")
+		if address == "" {
+			cli.ShowSubcommandHelp(c)
+			return nil
+		}
+		name := c.String("name")
+		if name == "" {
+			rbuf := make([]byte, 4)
+			rand.Read(rbuf)
+			name = "DNA-" + hex.EncodeToString(rbuf)
+		}
+		//create transaction
+		var tx *transaction.Transaction
+
+		wallet := openWallet(c.String("wallet"), []byte(c.String("password")))
```

Comment:
@dreamfly281 c.String("wallet")c.String("password") mean parse the string value after the arg "wallet" and "password" and pass them to openwallet